File: frontend/types/imports.ts
import type { BulkRowError } from '~/types/masterData'

export interface ImportPreview {
  batch_id: string
  entity_type: string
  status: 'validated' | 'invalid'
  mapping_profile: string
  mapping_version: string
  detected_columns: string[]
  applied_mapping: Record<string, string>
  total_rows: number
  valid_rows: number
  error_rows: number
  errors: BulkRowError[]
  sample: Record<string, unknown>[]
}

export interface ImportCommitResult {
  batch_id: string
  status: 'committed'
  imported_rows: number
}

export interface ImportErrorRecord {
  id: string
  row_number: number
  column_name: string | null
  error_code: string
  message: string
  raw_value: unknown | null
}

export interface ImportBatch {
  id: string
  entity_type: string
  filename: string
  mapping_profile: string
  mapping_version: string
  status: string
  total_rows: number
  valid_rows: number
  error_rows: number
  imported_rows: number
  created_at: string
  created_by: string | null
  errors: ImportErrorRecord[]
}

File: backend/app/schemas/imports.py
"""Excel import/export API schemas."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.master_data import BulkRowError


class ImportPreviewResponse(BaseModel):
    batch_id: UUID
    entity_type: str
    status: str
    mapping_profile: str
    mapping_version: str
    detected_columns: list[str]
    applied_mapping: dict[str, str]
    total_rows: int
    valid_rows: int
    error_rows: int
    errors: list[BulkRowError]
    sample: list[dict[str, Any]]


class ImportCommitRequest(BaseModel):
    batch_id: UUID


class ImportCommitResponse(BaseModel):
    batch_id: UUID
    status: str
    imported_rows: int


class ImportErrorRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    row_number: int
    column_name: str | None
    error_code: str
    message: str
    raw_value: Any | None


class ImportBatchRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    entity_type: str
    filename: str
    file_sha256: str
    mapping_profile: str
    mapping_version: str
    status: str
    total_rows: int
    valid_rows: int
    error_rows: int
    imported_rows: int
    created_at: datetime
    updated_at: datetime
    created_by: UUID | None
    errors: list[ImportErrorRead] = Field(default_factory=lambda: list[ImportErrorRead]())


class MappingOverride(BaseModel):
    source_to_target: dict[str, str] = Field(default_factory=dict)
